### PR TITLE
.Net: Handlebars planner complex types

### PR DESCRIPTION
**Pull request by teresaqhoang**
### Motivation and Context

Support for complex types in Handlebars Planner

### Description

Native Plugin with Complex Types
image

Remote Plugin with Complex Types
image

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/samples/KernelSyntaxExamples/Program.cs [hunk 1]
```diff
@@ -21,7 +21,7 @@ public static async Task Main(string[] args)
         using CancellationTokenSource cancellationTokenSource = new();
         CancellationToken cancelToken = cancellationTokenSource.ConsoleCancellationToken();
 
-        string? defaultFilter = null; // Modify to filter examples
+        string? defaultFilter = "Example66"; // Modify to filter examples
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Program.cs:24]: Revert this change

Comment (teresaqhoang) [reply dotnet/samples/KernelSyntaxExamples/Program.cs]: Done!

File: dotnet/samples/KernelSyntaxExamples/Example66_HandlebarsPlannerWithComplexTypes.cs [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.Functions.OpenAPI.Extensions;
+using Microsoft.SemanticKernel.Planners.Handlebars;
+using RepoUtils;
+
+/**
+ * This example shows how to use the Handlebars sequential planner with complex types.
+ */
+public static class Example66_HandlebarsPlannerWithComplexTypes
```
Comment (teresaqhoang) [inline dotnet/samples/KernelSyntaxExamples/Example66_HandlebarsPlannerWithComplexTypes.cs:20]: Currently working to merge this with Example65, had been waiting on the other PR to merge first

File: dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars [hunk 3]
```diff
@@ -2,19 +2,23 @@
 Explain how to achieve the user's goal with the available helpers with a Handlebars template.
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars:2]: The naming of this file can be improved, instead of `skPrompt` use a name that describes the purpose of the prompt. We won't to avoid having multiple files with the same name.

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs [hunk 4]
```diff
@@ -98,11 +101,55 @@ public async Task<HandlebarsPlan> CreatePlanAsync(string goal, CancellationToken
...
+        }
+
+        return functionsView;
+    }
+
+    // Extract any complex schemas for isolated render in prompt template
+    private ParameterView SetComplexTypeDefinition(ParameterView parameter)
+    {
+        // TODO (@teresaqhoang): handle case when schema and ParameterType can exist i.e., when ParameterType = RestApiResponse
+        if (parameter.ParameterType != null)
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:136]: Use `is not`

File: dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs [hunk 5]
```diff
@@ -69,9 +89,15 @@ private static async Task RunSampleAsync(string goal, params string[] pluginDire
         }
 
         // The gpt-35-turbo model does not handle loops well in the plans.
-        var allowLoopsInPlan = chatDeploymentName.Contains("gpt-35-turbo", StringComparison.OrdinalIgnoreCase) ? false : true;
+        var allowLoopsInPlan = !chatDeploymentName.Contains("gpt-35-turbo", StringComparison.OrdinalIgnoreCase);
```
Comment (TaoChenOSU) [inline dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs:92]: nit: 
1. Be nice if this explicitly mentions what models to use if loops are desired. Will gpt-4 work better, or it needs to be gpt4-turbo?
2. What is expected to happen if a goal requires loops but the planner is using 35 turbo?

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs [hunk 6]
```diff
@@ -54,55 +58,108 @@ public HandlebarsPlanner(IKernel kernel, HandlebarsPlannerConfig? config = defau
...
 
     private List<FunctionView> GetAvailableFunctionsManual(CancellationToken cancellationToken = default)
     {
-        return this._kernel.Functions.GetFunctionViews()
+        var availableFunctions = this._kernel.Functions.GetFunctionViews()
             .Where(s => !this._config.ExcludedPlugins.Contains(s.PluginName, StringComparer.OrdinalIgnoreCase)
                 && !this._config.ExcludedFunctions.Contains(s.Name, StringComparer.OrdinalIgnoreCase)
                 && !s.Name.Contains("Planner_Excluded"))
-            .ToList();
+                .ToList();
```
Comment (TaoChenOSU) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:111]: nit: is this extra tab necessary?

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs [hunk 7]
```diff
@@ -54,55 +58,108 @@ public HandlebarsPlanner(IKernel kernel, HandlebarsPlannerConfig? config = defau
...
+        if (parameter.ParameterType is not null)
+        {
+            // Async return type - need to extract the actual return type and override ParameterType property
+            var type = parameter.ParameterType;
+            if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Task<>))
+            {
+                parameter = parameter with { ParameterType = type.GenericTypeArguments[0] }; // Actual Return Type
+            }
+
+            this._parametersTypeView.UnionWith(parameter.ParameterType.ToHandlebarsParameterTypeView());
```
Comment (TaoChenOSU) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:148]: With this class member, it looks like the planner object of this class is expected to be discarded after a plan is created. Is this correct?

File: dotnet/samples/KernelSyntaxExamples/Example66_HandlebarsPlannerWithComplexTypes.cs (hunk 2 above)
Comment (teresaqhoang) [reply dotnet/samples/KernelSyntaxExamples/Example66_HandlebarsPlannerWithComplexTypes.cs]: Merged

File: dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars (hunk 3 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars:2]: Changed to `CreatePlanPrompt.handlebars`

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs (hunk 4 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs]: Changes in all instances

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs (hunk 6 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs]: nope, fixed now!

File: dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs (hunk 5 above)
Comment (teresaqhoang) [reply dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs:93]: It still works w 35-turbo just not as well

Changed comments to:
```
// Use gpt-4 or newer models if you want to test with loops. 
// Older models like gpt-35-turbo are able handle loops but is more prone to syntax errors.
```

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs (hunk 7 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:148]: Yep! It's only needed for prompt rendering

Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs]: Synced offline -- removed complex type definitions as class properties. These views now only exist in request scope of CreatePlanAsync

**Review by TaoChenOSU — APPROVED** [review]

File: dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs [hunk 8]
```diff
@@ -3,42 +3,51 @@
...
-        await RunDictionarySampleAsync();
+        await RunDictionaryWithBasicTypesSampleAsync();
         await RunPoetrySampleAsync();
         await RunBookSampleAsync();
+
+        // Using Complex Types as inputs and outputs
+        await RunLocalDictionaryWithComplexTypesSampleAsync(shouldPrintPrompt);
+
+        // TODO (@teresaqhoang): Enable this sample once complex types are supported for remote plugins return type.
+        // await RunRemoteDictionaryWithComplexTypesSampleAsync();
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs:42]: Can we leave this out until it's working?

File: dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs [hunk 9]
```diff
@@ -59,9 +68,20 @@ private static async Task RunSampleAsync(string goal, params string[] pluginDire
...
 
-        if (pluginDirectoryNames[0] == DictionaryPlugin.PluginName)
+        if (pluginDirectoryNames[0] == BasicDictionaryPlugin.PluginName)
+        {
+            kernel.ImportPluginFromObject(new BasicDictionaryPlugin(), BasicDictionaryPlugin.PluginName);
+        }
+        else if (pluginDirectoryNames[0] == ComplexDictionaryPlugin.PluginName)
+        {
+            kernel.ImportPluginFromObject(new ComplexDictionaryPlugin(), ComplexDictionaryPlugin.PluginName);
+        }
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs:78]: The naming of the two dictionary plugins isn't super obvious, something like the following would be clearer

1. `StringParamsDictionaryPlugin`
1. `ComplexParamsDictionaryPlugin`

File: dotnet/src/Planners/Planners.Core/Handlebars/CreatePlanPrompt.handlebars [hunk 10]
```diff
@@ -127,7 +182,7 @@ IMPORTANT: You can only use the helpers that are listed above. Do not use any ot
 - Do not make up values. Use the helpers to generate the data you need or extract it from the goal.
 - Keep data well-defined. Each variable should have a unique name. Create and assign each variable only once.
 - Be extremely careful about types. For example, if you pass an array to a helper that expects a number, the template will error out.
-- Avoid using loops. Try a solution without loops before you deploy a loop.
+{{#if allowLoops}}- Avoid using loops. Try a solution without before you deploy a loop.{{/if}}
 - There is no need to check your results in the template.
 - Do not nest sub-expressions or helpers because it will cause the template to error out.
 - Each step should contain only one helper call.
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/CreatePlanPrompt.handlebars:188]: We definitely need to figure out how to make this more maintainable

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/ParameterViewExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        // Add nested complex types
+        foreach (var property in properties)
+        {
+            parameterTypes.UnionWith(property.PropertyType.ToHandlebarsParameterTypeView());
+        }
+    }
+
+    public static bool isPrimitiveOrStringType(string type) =>
+        type == "string" || type == "number" || type == "integer" || type == "boolean" || type == "null";
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/Extensions/ParameterViewExtensions.cs:80]: Looking below `null` is an object and not a string or primitive, so is the above correct?

File: dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterView.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.SemanticKernel.Planners.Handlebars.Models;
+
+internal class HandlebarsParameterTypeView
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterView.cs:8]: We are changing our naming from `View` to `Metadata`, see https://github.com/microsoft/semantic-kernel/pull/3545

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/ParameterViewExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.Planners.Handlebars.Models;
+
+namespace Microsoft.SemanticKernel.Planners.Handlebars.Extensions;
+
+internal static class ParameterViewExtensions
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/Extensions/ParameterViewExtensions.cs:13]: Changing naming from `View` to `Metadata` see https://github.com/microsoft/semantic-kernel/pull/3545

**Review by markwallace-microsoft — COMMENTED** [review]
There is a rename coming which this PR should adopt

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/ParameterViewExtensions.cs (hunk 13 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Extensions/ParameterViewExtensions.cs]: Refactored everything from `View` -> `Metadata`

File: dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterView.cs (hunk 12 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterView.cs]: Refactored everything from `View` -> `Metadata`

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/ParameterViewExtensions.cs (hunk 11 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Extensions/ParameterViewExtensions.cs]: You're right, I moved the "null" type check from here to the conditional below

File: dotnet/src/Planners/Planners.Core/Handlebars/CreatePlanPrompt.handlebars (hunk 10 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/CreatePlanPrompt.handlebars:190]: Yeah, I'm thinking we can split it with partials, and it'll be more digestible.

File: dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs (hunk 9 above)
Comment (teresaqhoang) [reply dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs:74]: Done

File: dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs (hunk 8 above)
Comment (teresaqhoang) [reply dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs]: Done

File: dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    public override bool Equals(object obj)
+    {
+        // Check if the obj is null or of a different type
+        if (obj == null || obj.GetType() != this.GetType())
+        {
+            return false;
+        }
+
+        // Cast the obj to HandlebarsParameterTypeMetadata
+        var other = (HandlebarsParameterTypeMetadata)obj;
```
Comment (SergeyMenshykh) [inline dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs:34]:
```suggestion
        if (obj is not HandlebarsParameterTypeMetadata other)
        {
            return false;
        }
```

File: dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+        for (int i = 0; i < list1.Count; i++)
+        {
+            if (!list1[i].Name.Equals(list2[i].Name, System.StringComparison.Ordinal) || !list1[i].ParameterType!.Equals(list2[i].ParameterType))
+            {
+                return false;
+            }
+        }
+
+        // If all elements are equal, return true
+        return true;
```
Comment (SergeyMenshykh) [inline dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs:65]: The code can be replaced by `return this.Properties.SequenceEqual(other.Properties);` if SKParameterMetadata class implements Equals method and if sequential comparison is what we need here.

File: dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    // A helper method to compute the hash code of a list of SKParameterMetadata
+    private static int GetPropertiesHashCode(List<SKParameterMetadata> list)
+    {
+        // Use a prime number to combine the hash codes of the elements
+        int hash = 19;
+
+        // Use the default SKParameterMetadata hash code for each element
+        foreach (var item in list)
+        {
+            hash = (hash * 37) + GetPropertyHashCode(item);
```
Comment (SergeyMenshykh) [inline dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs:95]: No need in this method if SKParameterMetadata implements GetHashCode method. @stephentoub Is there an easier way in .NET to perform hash calculations?

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+#pragma warning disable IDE0130 // Namespace does not match folder structure
+namespace Microsoft.SemanticKernel.Planning.Handlebars.Extensions;
+#pragma warning restore IDE0130
+
+internal static class SKParameterMetadataExtensions
+{
+    /// <summary>
+    /// Checks if type is primitive or string
+    /// </summary>
+    public static bool isPrimitiveOrStringType(Type type) => type.IsPrimitive || type == typeof(string);
```
Comment (SergeyMenshykh) [inline dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs:20]:
```suggestion
    public static bool IsPrimitiveOrStringType(Type type) => type.IsPrimitive || type == typeof(string);
```

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    private static void AddNestedComplexTypes(this HashSet<HandlebarsParameterTypeMetadata> parameterTypes, PropertyInfo[] properties)
+    {
+        // Add nested complex types
+        foreach (var property in properties)
+        {
+            parameterTypes.UnionWith(property.PropertyType.ToHandlebarsParameterTypeMetadata());
+        }
+    }
+
+    public static bool isPrimitiveOrStringType(string type) =>
```
Comment (SergeyMenshykh) [inline dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs:81]:
```suggestion
    public static bool IsPrimitiveOrStringType(string type) =>
```

Comment (SergeyMenshykh) [inline dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs:81]: Do we need this overload of the `isPrimitiveOrStringType` method taking into account we already declared one above? 
Can we keep the two methods next to each other?

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlan.cs [hunk 19]
```diff
@@ -14,15 +16,23 @@ public sealed class HandlebarsPlan
...
     /// <summary>
     /// Initializes a new instance of the <see cref="HandlebarsPlan"/> class.
     /// </summary>
-    /// <param name="kernel">The kernel.</param>
-    /// <param name="template">The Handlebars template.</param>
-    public HandlebarsPlan(Kernel kernel, string template)
+    /// <param name="kernel">Kernel instance.</param>
+    /// <param name="template">A Handlebars template representing the plan.</param>
+    /// <param name="createPlanPromptTemplate">Prompt template used to generate the plan.</param>
+    public HandlebarsPlan(Kernel kernel, string template, string createPlanPromptTemplate)
```
Comment (SergeyMenshykh) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlan.cs:31]: Consider adding more details about `template` and `createPlanPromptTemplate` parameters to clearly describe the difference between the two.

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs [hunk 20]
```diff
@@ -66,11 +69,11 @@ internal sealed class HandlebarsTemplateEngineExtensions
...
         IHandlebars handlebarsInstance,
-        SKFunctionMetadata functionView,
+        SKFunctionMetadata functionMetadata,
         Dictionary<string, object?> variables,
         CancellationToken cancellationToken = default)
     {
-        string fullyResolvedFunctionName = functionView.PluginName + ReservedNameDelimiter + functionView.Name;
+        string fullyResolvedFunctionName = functionMetadata.PluginName + ReservedNameDelimiter + functionMetadata.Name;
 
         handlebarsInstance.RegisterHelper(fullyResolvedFunctionName, (in HelperOptions options, in Context context, in Arguments arguments) =>
```
Comment (SergeyMenshykh) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs:78]: Consider splitting the `RegisterHelper` callback in a few smaller private methods to it readable.

File: dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs (hunk 15 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs:65]: We do need sequential property comparison but only need to check Name and PropertyType from the SKParameterMetadata object, since those are the only 2 properties we use in the HandlebarsParameterTypeMetadata object, 
image

Also, SKParameterMetadata doesn't override the Equals method.

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs (hunk 18 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs:81]: Yes one takes a `Type` parameter and the other takes a stringified type (type serialized in JSON Schema). Moved them to be next to each other

image

File: dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs (hunk 15 above)
Comment (stephentoub) [reply dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs:65]: > Also, SKParameterMetadata doesn't override the Equals method.

We could make it do so if desirable.

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlan.cs (hunk 19 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlan.cs:31]: I added some comments as well as changed the parameter name in the constructor:
image

I didn't want to change the class `_template` property name because ``HandlebarsPlan._template` makes sense

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs (hunk 20 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs:78]: Done, split into `ProcessHashArguments`, `ProcessPositionalArguments`, `InitializeContextVariables`, and `InvokeSKFunction`

File: dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs (hunk 15 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs:62]: This scenario is pretty specific to the Planners and how we render complex types for the function manual, but I think it could be beneficial to centralize this in the SKParameterMetadata object anyway (for future planners or data checks when sharing variables across multiple functions of the same plugin)

File: dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs (hunk 16 above)
Comment (stephentoub) [reply dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs:92]: System.HashCode is the recommended way to combine multiple hash codes into a single one, e.g.
```C#
public override int GetHashCode()
{
    HashCode hc = default;
    hc.Add(this.Name);
    hc.Add(this.IsComplex);
    foreach (var item in this.Properties)
    {
        hc.Add(item?.Name);
        hc.Add(item?.ParameterType);
    }
    return hc.ToHashCode();
}
```
That said, also keep in mind that while equality does need to factor in all relevant data, hash code doesn't have to. It just needs to factor in enough to provide a reasonably good-enough distribution.  You could, for example, just have the implementation be:
```C#
public override int GetHashCode() => this.Name?.GetHashCode() ?? 0;
```
and depending on the circumstances, that might be "good enough".

Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Models/HandlebarsParameterTypeMetadata.cs]: Refactored to use HashCode

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/HandlebarsPlannerExtensions.cs [hunk 21]
```diff
@@ -4,7 +4,9 @@
 using System.IO;
 using System.Reflection;
 
-namespace Microsoft.SemanticKernel.Planning.Handlebars;
+#pragma warning disable IDE0130 // Namespace does not match folder structure
+namespace Microsoft.SemanticKernel.Planning.Handlebars.Extensions;
+#pragma warning restore IDE0130
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/Extensions/HandlebarsPlannerExtensions.cs:9]: I'd expect the extensions to be in th same namespace as the class they are extending

Nit: there is no need to do warning restore here

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.Planning.Handlebars.Models;
+
+#pragma warning disable IDE0130 // Namespace does not match folder structure
+namespace Microsoft.SemanticKernel.Planning.Handlebars.Extensions;
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs:12]: Is this the best namespace?

**Review by markwallace-microsoft — APPROVED** [review]
Couple of minor things.

Next PR we should move this to it's own package.

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/HandlebarsPlannerExtensions.cs (hunk 21 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Extensions/HandlebarsPlannerExtensions.cs:9]: This was an intentional change by Stephen to consolidate everything into the `Planning` namespace: https://github.com/microsoft/semantic-kernel/pull/3564

And I see this error if I remove the warning ignore:
image

Comment (stephentoub) [reply dotnet/src/Planners/Planners.Core/Handlebars/Extensions/HandlebarsPlannerExtensions.cs:9]: But why is `.Extensions` needed?

Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Extensions/HandlebarsPlannerExtensions.cs:9]: Ah I see what you mean, I removed it

File: dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs (hunk 22 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs]: Removed `.Extensions`, just `namespace Microsoft.SemanticKernel.Planning.Handlebars` now

**Review by markwallace-microsoft — APPROVED** [review]

**Review by TaoChenOSU — APPROVED** [review]

**Review by TaoChenOSU — APPROVED** [review]
